cmd/collector/app/span_processor.go

```diff
@@ -177,3 +206,76 @@ func (sp *spanProcessor) enqueueSpan(span *model.Span, originalFormat SpanFormat
 	}
 	return sp.queue.Produce(item)
 }
+
+func (sp *spanProcessor) StartStateReporting(reportPeriod time.Duration) {
+	sp.background(reportPeriod, func() {
+		sp.updateGauges()
+	})
+}
+
+func (sp *spanProcessor) StartDynQueueSizeUpdater(reportPeriod time.Duration) {
+	sp.background(reportPeriod, func() {
+		sp.updateQueueSize()
+	})
+}
+
+func (sp *spanProcessor) background(reportPeriod time.Duration, callback func()) {
+	ticker := time.NewTicker(reportPeriod)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				callback()
+			case <-sp.stopCh:
+				return
+			}
+		}
+	}()
+}
+
+func (sp *spanProcessor) updateQueueSize() {
+	if sp.spansProcessed == 0 {
+		return
+	}
+
+	if sp.dynQueueSizeWarmup == 0 {
+		return
+	}
+
+	if sp.spansProcessed < uint64(sp.dynQueueSizeWarmup) {
+		return
+	}
+
+	// first, we get the average size of a span, by dividing the bytes processed by num of spans
+	average := sp.bytesProcessed / sp.spansProcessed
+
+	// finally, we divide the available memory by the average size of a span
+	idealQueueSize := float64(sp.dynQueueSizeMemory / uint(average))
+
+	// cap the queue size, just to be safe...
+	if idealQueueSize > maxQueueSize {
+		idealQueueSize = maxQueueSize
+	}
+
+	var diff float64
+	current := float64(sp.queue.Capacity())
+	if idealQueueSize > current {
+		diff = idealQueueSize / current
+	} else {
+		diff = current / idealQueueSize
+	}
+
+	// resizing is a costly operation, we only perform it if we are at least 10% apart from the desired value
+	if diff > 1.1 {
+		s := int(idealQueueSize)
+		sp.logger.Info("Resizing the internal span queue", zap.Int("new-size", s), zap.Uint64("average-span-size-bytes", average))
+		sp.queue.Resize(s)
```

Comment:
I think there needs to be a mutex to prevent concurrent resize attempts

Reply:
There should be only one such operation per minute at most, but you are right. Better be safe here.